pkg/kubeadm/cmd/init.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/renstrom/dedent"
+	"github.com/spf13/cobra"
+
+	kubeadmapi "k8s.io/kubernetes/pkg/kubeadm/api"
+	kubemaster "k8s.io/kubernetes/pkg/kubeadm/master"
+	kubeadmutil "k8s.io/kubernetes/pkg/kubeadm/util"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+var (
+	init_done_msgf = dedent.Dedent(`
+		Kubernetes master initialised successfully!
+
+		You can connect any number of nodes by running:
+
+		kubeadm join --token %s %s
+		`)
+)
+
+func NewCmdInit(out io.Writer, params *kubeadmapi.BootstrapParams) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "init --token <secret> [--listen-ip <addr>]",
+		Short: "Run this on the first server you deploy onto.",
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunInit(out, cmd, args, params)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&params.Discovery.ListenIP, "listen-ip", "", "",
+		`(optional) IP address to listen on, in case autodetection fails.`)
+	cmd.PersistentFlags().StringVarP(&params.Discovery.GivenToken, "token", "", "",
+		`(optional) Shared secret used to secure bootstrap. Will be generated and displayed if not provided.`)
+
+	return cmd
+}
+
+func RunInit(out io.Writer, cmd *cobra.Command, args []string, params *kubeadmapi.BootstrapParams) error {
+	if params.Discovery.ListenIP == "" {
+		ip, err := kubeadmutil.GetDefaultHostIP()
+		if err != nil {
+			return err
+		}
+		params.Discovery.ListenIP = ip
+	}
+	if err := kubemaster.CreateTokenAuthFile(params); err != nil {
+		return err
+	}
+	if err := kubemaster.WriteStaticPodManifests(params); err != nil {
+		return err
+	}
+	caKey, caCert, err := kubemaster.CreatePKIAssets(params)
+	if err != nil {
+		return err
+	}
+	kubeconfigs, err := kubemaster.CreateCertsAndConfigForClients(params, []string{"kubelet", "admin"}, caKey, caCert)
```

Comment:
Yes, it'd be surprising to find a hiden file in root's home... So I left it
in etc. This is meant to be a recovery file of sorts, if all goes nuts. The
idea is to get user certs using CSR API, and use email address as CN etc.

On Wed, 31 Aug 2016, 22:39 Joe Beda, <notifications@github.com> wrote:

> In pkg/kubeadm/cmd/init.go
> <https://github.com/kubernetes/kubernetes/pull/31221#discussion_r77078655>
> :
>
> > +		if err != nil {
> > +			return err
> > +		}
> > +		params.Discovery.ListenIP = ip
> > +	}
> > +	if err := kubemaster.CreateTokenAuthFile(params); err != nil {
> > +		return err
> > +	}
> > +	if err := kubemaster.WriteStaticPodManifests(params); err != nil {
> > +		return err
> > +	}
> > +	caKey, caCert, err := kubemaster.CreatePKIAssets(params)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	kubeconfigs, err := kubemaster.CreateCertsAndConfigForClients(params, []string{"kubelet", "admin"}, caKey, caCert)
>
> It is a little wacky here that we are creating the admin key and writing
> it in to the system. It's probably okay for now...
>
> Ideas:
>
>    - Don't create admin creds and do something with a unix domain socket.
>    - Write out some sort of file that you can import and pass around.
>    Requires new support in kubectl.
>    - Write it in to the user's home directory instead of a system wide
>    directory. However, because of the need to write static pod manifests most
>    likely we'll be running as root and won't know the right "user" to write it
>    to.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31221/files/57096ca902c5c666d45df191037d9278972af1fd#r77078655>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS7VM7odlzc917SztsrUmYhylzrk2ks5qlfSDgaJpZM4JqsTH>
> .
>